### PR TITLE
Feat/delete users tokens

### PR DESCRIPTION
**Pull request by timDeHof**
# Description

This PR adds a DELETE endpoint to auth services that will revoke an user's refresh token when given a user's id or email.This endpoint has role guard attached so that only users with admin role can initiate the endpoint.  

Also this PR tests the new functionality of revoking a refresh token. The tests cover scenarios where the refresh token is successfully revoked, where the email is invalid, and where the user is not permitted to revoke the token. These tests ensure that the revoking of refresh tokens is working correctly and that the appropriate responses are returned based on the different scenarios.

In addition, swagger UI api has been updated to allow for users to test the new endpoint.

## Issue link

Fixes # (issue) "Create endpoint to revoke refresh token" from trello

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Feature updates / changes
- [ ] Tests
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

You can run `yarn test:e2e` or `yarn test:e2e:docker`, then look for `auth.e2e-spec.ts` to pass.
Screenshot 2024-03-08 at 1 05 13 PM
Screenshot 2024-03-08 at 1 04 41 PM

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have updated the change log

### CODE REVIEW

File: src/auth/auth.controller.ts [hunk 1]
```diff
@@ -196,6 +202,38 @@ export class AuthController {
...
+        type: BadRequestErrorResponse,
+    })
+    @ApiResponse({
+        status: HttpStatus.FORBIDDEN,
+        description: "user doesn't have permission to preform operation",
+        type: ForbiddenErrorResponse,
+    })
+    @HttpCode(HttpStatus.OK)
+    @Roles(AppRoles.Admin)
+    @UseGuards(RolesGuard)
```
Comment (cherylli) [inline src/auth/auth.controller.ts:231]: @useGuards(RolesGuard) shouldn't be needed, as we have a app level rolesGuard 
(`src/app.module.ts`)

File: src/auth/auth.controller.ts [hunk 2]
```diff
@@ -34,6 +36,10 @@ import { JwtAuthGuard } from "./guards/jwt-auth.guard";
 import { JwtRefreshAuthGuard } from "./guards/jwt-rt-auth.guard";
 import { Public } from "../global/decorators/public.decorator";
 import { AT_MAX_AGE, RT_MAX_AGE } from "../global/constants";
+import { RevokeRTDTo } from "./dto/revoke-refresh-token.dto";
```
Comment (cherylli) [inline src/auth/auth.controller.ts:39]:
```suggestion
import { RevokeRTDto } from "./dto/revoke-refresh-token.dto";
```

File: test/auth.e2e-spec.ts [hunk 3]
```diff
@@ -421,6 +421,86 @@ describe("AuthController e2e Tests", () => {
...
+                data: {
+                    refreshToken: null,
+                },
+            });
+
+            await request(app.getHttpServer())
+                .delete(revokeRTUrl)
+                .set("Cookie", [access_token, refresh_token])
+                .send({ email: "l.castro@outlook.com" })
+                .expect(200);
```
Comment (cherylli) [inline test/auth.e2e-spec.ts:447]: should we add an expect (refreshtoken tobe null) in the database after revoke?

File: test/auth.e2e-spec.ts [hunk 4]
```diff
@@ -421,6 +421,86 @@ describe("AuthController e2e Tests", () => {
...
+                .delete(revokeRTUrl)
+                .set("Cookie", [access_token, refresh_token])
+                .send({
+                    userId: "c4daf07c-8dde-4a43-bfa4-a6fd49762dd5",
+                    email: "l.castro@outlook.com",
+                })
+                .expect(400);
+        });
+
+        it("should return 403 if user is not permitted", async () => {
```
Comment (cherylli) [inline test/auth.e2e-spec.ts:485]: I like that you included this test, we should definitely go back to the older tests to include these once we added other permissions

**Review by cherylli — CHANGES_REQUESTED** [review]
this PR is very well done. Thank you

File: test/auth.e2e-spec.ts (hunk 3 above)
Comment (timDeHof) [reply test/auth.e2e-spec.ts:447]: That would be a good idea. But when I tried to add `.expect(refreshToken).toBeNull()`, I got an error saying "Property 'toBeNull' does not exist on type 'Test'"

Comment (timDeHof) [reply test/auth.e2e-spec.ts:447]: I added a separate test that checks that the user's refresh token in null

**Comment by cherylli** [comment]
image
I think this is not accurate, since we can either use email or userID, maybe it should be `auth/refresh/revoke` or `auth/revoke-refresh-token`

You can look at this for reference https://restfulapi.net/resource-naming/

------------------
image
I got this when user is not found in the database, it's not very descriptive

------------------
200 - delete token successful works, but there's no response message
image

**Review by cherylli — CHANGES_REQUESTED** [review]
image
I think this is not accurate, since we can either use email or userID, maybe it should be `auth/refresh/revoke` or `auth/revoke-refresh-token`

You can look at this for reference https://restfulapi.net/resource-naming/

------------------
image
I got this when user is not found in the database, it's not very descriptive

------------------
200 - delete token successful works, but there's no response message
image

File: src/auth/auth.service.ts [hunk 5]
```diff
@@ -133,6 +135,54 @@ export class AuthService {
...
+                        id: userId,
+                    },
+                    data: {
+                        refreshToken: null,
+                    },
+                });
+            } catch (error) {
+                throw new NotFoundException({
+                    statusCode: 404,
+                    message: `User by "${userId}" cannot be found.`,
```
Comment (cherylli) [inline src/auth/auth.service.ts:159]:
```suggestion
                    message: `User by '${userId}' cannot be found.`,
```

just look better on swagger 
image
vs
image

File: src/auth/auth.service.ts [hunk 6]
```diff
@@ -133,6 +135,54 @@ export class AuthService {
...
+                        email: email,
+                    },
+                    data: {
+                        refreshToken: null,
+                    },
+                });
+            } catch (error) {
+                throw new NotFoundException({
+                    statusCode: 404,
+                    message: `User by "${email}" cannot be found.`,
```
Comment (cherylli) [inline src/auth/auth.service.ts:175]:
```suggestion
                    message: `User by '${email}' cannot be found.`,
```

**Review by cherylli — APPROVED** [review]
thanks for the changes!

**Review by cherylli — APPROVED** [review]

**Review by JoshuaHinman — APPROVED** [review]
e2e tests passed , swagger responses work, code looks great!

**Pull request merged**